src/toolchain.ml

```diff
@@ -181,14 +181,73 @@ let of_settings () : t option Promise.t =
           (Opam.Switch.name switch);
         None
       ) )
-  | Some Global -> Promise.return (Some Global)
+  | Some System -> Promise.return (Some System)
   | Some (Custom template) -> Promise.return (Some (Custom template))
 
+let detect_esy_sandbox ~project_root esy () =
+  let open Promise.Syntax in
+  let* esy = esy in
+  let* esy_build_dir_exists = Fs.exists (Filename.concat project_root "_esy") in
+  let* manifest = Esy.find_manifest_in_dir project_root in
+  match (esy, esy_build_dir_exists, manifest) with
+  | Some esy, true, Some manifest ->
+    (* Some projects have both use both Esy and Opam, we can't assume that the user
+       wants to use Esy just by looking for an "esy.json" file.
+
+       Instead, we assume the user wants to use esy if these three conditions are met:
+        - The esy command is in the user's system
+        - There is an _esy build directory in the project root
+        - There is an Esy manifest file in the project root
+    *)
+    Promise.return (Some (Esy (esy, Path.of_string manifest)))
+  | _ -> Promise.return None
+
+let detect_opam_sandbox ~project_root opam () =
+  let open Promise.Syntax in
+  let* opam = opam in
+  match opam with
+  | None -> Promise.return None
+  | Some opam -> (
+    let+ switch = Opam.switch_show ~cwd:project_root opam in
+    match switch with
+    | Some switch -> Some (Opam (opam, switch))
+    | None -> None )
+
+let detect () =
+  let rec aux = function
+    | [] -> Promise.return None
+    | el :: rest -> (
+      let open Promise.Syntax in
+      let* sandbox_opt = el () in
+      match sandbox_opt with
+      | Some x -> Promise.return (Some x)
+      | None -> aux rest )
```

Comment:
My preference is against polluting application code with generalized functions. So I would move this into `Promise.List.find_map` or some such.